### PR TITLE
Luis Review and Edits 

### PR DESCRIPTION
**Pull request by bytezantium**
I reviewed, edited and added.

The only thing I left out is the Withdrawal part. 

Review next time.

Next I will work on executor reward payout mechanism and submit a PR tomorrow midday.

### CODE REVIEW

**Comment by bytezantium** [comment]
@hilmarx  habe ein fehler bei mir gefunden:

have in createSellOrder() vergessen den lastParticipationIndex auf 0 zu defaulten, anstatt den da schon zu fetchen.

Kannste rausediten.

File: contracts/Gelato.sol [hunk 1]
```diff
@@ -1,234 +1,539 @@
...
+                    * We increment our lastParticipationIndex by 1.
+                       Compared to Case3, incrementing by 1 suffices in Case4
+                       because the lastParticipationIndex is not 1 behind the
+                       currentAuctionIndex, like it is in Case3.
+            * INTERACTIONS
+                    * we _depositAndSell()
+            * EVENT
+                * we emit LogSubOrderExecuted(sellOrderHash, seller, executor)
+        */
+        else if (currentAuctionIndex == sellOrder.lastParticipationIndex) {
```
Comment (hilmarx) [inline contracts/Gelato.sol]: First sell order won't execute if we are currently in the buffer period or waiting period, even though we have never invested before. Problem for illiquid tokens where our first subOrder will fill up the necessary $1k.

**Comment by hilmarx** [comment]
Summary:

1) Next time editing files: As you used exactly the same conditions (for Case 2 - 4 ), it would make things a lot easier if you change them in the order it which they were. It took an hour to realize that they are the same just in a different order. Next time please make a PR with changing the functions and then in another one re ordering them. 

2) Bug in logic. Assume an order is created with Auction index (AI) = 10. It is currently waiting for funding because the token is illiquid. Now we want to execute the order because we are sure we can fill it in order to start the auction. Case 4 would be responsible for that. However, as we are in the waiting for funding period, it wont trigger. Hence we will only execute the transaction if other parties fill up the auction and it starts eventually.

3) For the conditions, I prefer thinking about newAuctionIndex == lastAuctionIndex + something rather than newAuctionIndex - something == lastAuctionIndex. More intuitive for me, is it really easier for you with subtracting the new?

4) I would stick to defaulting the lastAuctionIndex to 0. This removes a lot of complexity inherent to the DutchX and makes testing a lot easier, because we always simply set the lastAuctionIndex = newAuctionIndex. In your cases, we do:

Case 4: lastAuctionIndex = 1 + lastAuctionIndex
Case 3: lastAuctionindex = 1 + currentAuctionIndex
Case 2: lastAuctionIndex = currentAuction Index

=> Lets remove complexity and stick to the same method for each of them. Reentrancy can also simply be avoided by incrementing the hammerTime before we sell.

**Comment by bytezantium** [comment]
General best practice: Let's indicate if we tested our claims and how/where we tested them, so that we can understand, test and reproduce.

File: contracts/Gelato.sol [hunk 2]
```diff
@@ -1,234 +1,539 @@
...
+        // Execute if: It's hammerTime !
+        require(subOrder.hammerTime >= now,
+            "Failed: You called before scheduled execution time"
+        );
+
+        // Execute if: Seller has the balance.
+        uint256 balance = ERC20(subOrder.sellToken).balanceOf(subOrder.seller);
+        // Question: Operator precedence in . chains
+        require(
+            balance >= subOrder.subOrderSize.add(subOrder.executionReward),
```
[the remark this answers is not in the document]
Comment (bytezantium) [reply contracts/Gelato.sol]: 1: executionReward is per subOrder. It has to be per subOrder (not totalBounty) because different executors might execute subOrders that form one sellOrder.

[the remark this answers is not in the document]
Comment (bytezantium) [reply contracts/Gelato.sol]: 2:
```suggestion
            balance >= subOrder.subOrderSize,
```

File: contracts/Gelato.sol [hunk 3]
```diff
@@ -1,234 +1,539 @@
...
-        SellOrder memory sellOrder = sellOrders[structHash];
+        // Argument checks
+        require(msg.sender != address(0), "No zero addresses allowed");
+        require(buyToken != address(0), "No zero addresses allowed");
+        require(sellToken != address(0), "No zero addresses allowed");
+        require(totalSellVolume != 0, "Empty sell volume");
+        require(subOrderSize != 0, "Empty sub order size");
+
+        // Invariant checks
+        require(numSubOrders == totalSellVolume.div(subOrderSize),
```
[the remark this answers is not in the document]
Comment (bytezantium) [reply contracts/Gelato.sol]: what change do you suggest to on-chain verify the correct subOrder on totalSellVolume mapping?

File: contracts/Gelato.sol [hunk 4]
```diff
@@ -1,234 +1,539 @@
...
+            balance >= subOrder.subOrderSize.add(subOrder.executionReward),
+            "Failed: Seller balance must be greater than subOrderSize + executionReward"
+        );
+
+        // Execute if: Gelato has the allowance.
+        uint256 allowance = ERC20(subOrder.sellToken)
+            .allowance(subOrder.seller, address(this)
+        );
+        require(
+            allowance >= subOrder.subOrderSize.add(executionReward),
```
[the remark this answers is not in the document]
Comment (bytezantium) [reply contracts/Gelato.sol]: btw this PR is pre payment logic. This is based off the ERC20 logic I branched from.  I am currently working on branch feature/flatEtherFee where I introduce the reward payment logic and where I delete all of that.

File: contracts/Gelato.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (bytezantium) [reply contracts/Gelato.sol]: 2: btw this PR is pre payment logic. This is based off the ERC20 logic I branched from.  I am currently working on branch feature/flatEtherFee where I introduce the reward payment logic and where I delete all of that.

File: contracts/Gelato.sol [hunk 5]
```diff
@@ -1,234 +1,539 @@
...
-            {
-                 // Continue to depositAndSell as long as other tests pass
-            }
+        uint256 auctionStartTime = DutchX.getAuctionStart(subOrder.sellToken,
+                                                          subOrder.buyToken
+        );
+
+        // Local variables for readability
+        bool noConsecutiveParticipation = (
+            sellOrder.lastParticipationIndex <= currentAuctionIndex.sub(2)
```
[the remark this answers is not in the document]
Comment (bytezantium) [reply contracts/Gelato.sol]: In my opinion we should think from the perspective of where we last participated. 
Mine says:
If we participated 2 or more auctions ago in the past. We are good to go.

Yours says:
If the current auction index on the DutchX is more than 1 ahead of the last auction we participated in, if we were to add 1 to our last participation index. We are good to go.

Or try to rephrase it better in English to reflect what your code says.

File: contracts/Gelato.sol [hunk 6]
```diff
@@ -1,234 +1,539 @@
...
+        );
+        bool waitingPeriod = (
+            AUCTION_START_WAITING_FOR_FUNDING || now < auctionStartTime
```
[the remark this answers is not in the document]
Comment (bytezantium) [reply contracts/Gelato.sol]: In 24 hour waiting case:

- we fire scheduleNextAuction after 24hours
- this sets a new startTime = now + WAITIN_PERIOD_NEW_AUCTION (10 minutes)
- now we still MUST wait 10 minutes before auctionStartTime.
- hence we must now wait until auctionStartTime < now.
- hence we are still in the same waitingPeriod as before.
- only after 10 minutes is auctionStart < now in postSellOrder
- only then do our funds go to sellVolumesNext
- we cannot have them go to sellVolumesCurrent 

In our logic there is no difference between AUCTION_START_WAITING_FOR_FUNDING and the consecutive BUFFER_PERIOD.

In our logic these 2 periods form 1 -> the WAITING_PERIOD.

This is due to the logic in DutchX postSellOrder()

Where funds only go to sellVolumesNext after our WAITING_PERIOD logic.

Please read through postSellOrder() carefully and scheduleNextAuction(). Let me know if/what you see differently in those functions, or elsewhere.

File: contracts/Gelato.sol [hunk 7]
```diff
@@ -1,234 +1,539 @@
...
-        } 
-        
-        // The current auction index is greater than the last one we sold into + 1 => We are safe
-        else if (newAuctionIndex > sellOrder.lastAuctionIndex.add(1) ) 
-        
-        {
-            // No checks required, continue to depositAndSell
+            else if (waitingPeriodOver) {
+                // EFFECTS
+                sellOrder.lastParticipationIndex++;
```
Comment (hilmarx) [inline contracts/Gelato.sol]: I get why you want to increment lastParticipationIndex here, to mimic the DutchX to avoid re-entrance. However we have other checks to avoid that such as freeze time. 

The only time we should actually sell in this scenario is:
- if we sold in the last waiting for funding or buffer period and Auction Start < now. However in our current implementation, we don't track in what specific period we sold, we only track the auction index. 

Hence the options are:

1) Playing it safe and increment the lastParicipationIndex by 1,
2) dont have index protection and rely on freeze time as require
3) Introduce another state variable that checks in which period we sold the previous auction.

If would be up to increment index by 1 here, but I am still not 100% sure about all the edge cases here.

File: contracts/Gelato.sol [hunk 8]
```diff
@@ -1,234 +1,539 @@
...
+
+        // Local variables for readability
+        bool noConsecutiveParticipation = (
+            sellOrder.lastParticipationIndex <= currentAuctionIndex.sub(2)
+        );
+        bool waitingPeriod = (
+            AUCTION_START_WAITING_FOR_FUNDING || now < auctionStartTime
+        );
+        // Same values but different context readability
+        bool untouchedOngoingAuction = auctionStartTime < now;
```
[the remark this answers is not in the document]
Comment (bytezantium) [reply contracts/Gelato.sol]: ok!

File: contracts/Gelato.sol [hunk 9]
```diff
@@ -1,234 +1,539 @@
...
+                      one to follow the DutchX rule that, if we invest into an ongoing auction,
+                      our auction index must be currentAuctionIndex + 1 (that
+                      is where the second + comes from. @Hilmar: this is the new stuff)
+                      -> total: lastParticipationIndex + 2 or lastParticipationIndex.add(2)
+            * INTERACTIONS
+                    * we _depositAndSell()
+            * EVENT
+                * we emit LogSubOrderExecuted(sellOrderHash, seller, executor)
+        */
+        else if (sellOrder.lastParticipationIndex == currentAuctionIndex.sub(1)) {
```
[the remark this answers is not in the document]
Comment (bytezantium) [reply contracts/Gelato.sol]: please write your code out in plain English to hear/see what is more intuitive.

Mine:
If we last participated in the previous auction index.

Yours:
If we were to add one to our last participation index, and that would be equal to the current auction index .....

File: contracts/Gelato.sol (hunk 8 above)
[the remark this answers is not in the document]
Comment (bytezantium) [reply contracts/Gelato.sol]: "untouched" because in the control flow where the variable is used, the logic dictates that we did not invest yet in the current auction that already started. This is more clear than "currentAuctionStarted" which leaves the reader to ponder whether we already invested in it or not.

File: contracts/Gelato.sol (hunk 1 above)
Comment (bytezantium) [reply contracts/Gelato.sol]: caution: in my language AUCTION_START_WAITING_FOR_FUNDING and WAITING_PERIOD_NEW_AUCTION ("buffer") together form the "waitingPeriod".
Why would first sell order not execute?

- First sell order is indexed 0
- First sell order should execure in Case 2 or Case 3.

no?

Comment (hilmarx) [reply contracts/Gelato.sol]: 1) Saw comment about bug fix with setting default auction index = 0 after this.

**Comment by bytezantium** [comment]
> Summary:
> 
> 1. Next time editing files: As you used exactly the same conditions (for Case 2 - 4 ), it would make things a lot easier if you change them in the order it which they were. It took an hour to realize that they are the same just in a different order. Next time please make a PR with changing the functions and then in another one re ordering them.
> 2. Bug in logic. Assume an order is created with Auction index (AI) = 10. It is currently waiting for funding because the token is illiquid. Now we want to execute the order because we are sure we can fill it in order to start the auction. Case 4 would be responsible for that. However, as we are in the waiting for funding period, it wont trigger. Hence we will only execute the transaction if other parties fill up the auction and it starts eventually.
> 3. For the conditions, I prefer thinking about newAuctionIndex == lastAuctionIndex + something rather than newAuctionIndex - something == lastAuctionIndex. More intuitive for me, is it really easier for you with subtracting the new?
> 4. I would stick to defaulting the lastAuctionIndex to 0. This removes a lot of complexity inherent to the DutchX and makes testing a lot easier, because we always simply set the lastAuctionIndex = newAuctionIndex. In your cases, we do:
> 
> Case 4: lastAuctionIndex = 1 + lastAuctionIndex
> Case 3: lastAuctionindex = 1 + currentAuctionIndex
> Case 2: lastAuctionIndex = currentAuction Index
> 
> => Lets remove complexity and stick to the same method for each of them. Reentrancy can also simply be avoided by incrementing the hammerTime before we sell.

File: contracts/Gelato.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (bytezantium) [reply contracts/Gelato.sol]: #35

File: contracts/Gelato.sol [hunk 10]
```diff
@@ -1,234 +1,539 @@
...
+        emit LogNewSellOrderCreated(sellOrderHash, seller, hammerTime);
+    }
 
-        // Check if sub-order is ready to be executed based on time
-        require(sellOrder.nextValidTimestamp > now);
-        
-        // ### Basic Checks End ###
 
-        // ### Advanced Checks ###
+    function executeSubOrder(bytes sellOrderHash)
```
Comment (hilmarx) [inline contracts/Gelato.sol]:
```suggestion
    function executeSubOrder(bytes32 sellOrderHash)
```

File: contracts/Gelato.sol [hunk 11]
```diff
@@ -1,234 +1,539 @@
...
+        require(!subOrder.cancelled,
+            "Failed: Sell Order has status cancelled."
+        );
+
+        // Execute if: It's hammerTime !
+        require(subOrder.hammerTime >= now,
+            "Failed: You called before scheduled execution time"
+        );
+
+        // Execute if: Seller has the balance.
```
Comment (hilmarx) [inline contracts/Gelato.sol]:
```suggestion
        // Execute if: Seller has the balance.
        // @DEV Revisit based on payout logic
```

File: contracts/Gelato.sol (hunk 2 above)
Comment (hilmarx) [inline contracts/Gelato.sol]:
```suggestion
            // @DEV revisit adding execution reward based on payout logic
            balance >= subOrder.subOrderSize.add(subOrder.executionReward),
```

File: contracts/Gelato.sol (hunk 7 above)
Comment (hilmarx) [reply contracts/Gelato.sol]: Postponed.

Comment (bytezantium) [reply contracts/Gelato.sol]: #36

File: contracts/Gelato.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (bytezantium) [reply contracts/Gelato.sol]: #10
